sstables: add SSTable compression with shared dictionaries

michoecho opened the pull request:
This PR extends Scylla's SSTable compression with the ability to use compression dictionaries shared across compression chunks. This involves several changes:

- We refactor `compression_parameters` and friends (`compressor`, `sstables::local_compression`, `sstables::compression`) to prepare for making the construction of `compressor`s asynchronous, to enable sharing pieces of compressors (the dictionaries) across shards.
- We introduce the notion of "hidden compression options" which are written to `CompressionInfo.db` and used to construct decompressors, like regular options, but don't appear in the schema. (We later stuff the SSTable's dictionary into `CompressionInfo.db` using a sequence of such options).
- We add a cluster feature which guards the creation of dictionary-compressed SSTables.
- We introduce a central "compressor factory" (one instance shared by all shards), which from this point onward is used to construct all `compressor` objects (one per SSTable) used to process the SSTables. When constructing a compressor for writing, it uses the "current"/"recommended" dictionary (which is passed to the factory from the actively-observed contents of the group0-managed `system.dicts`). When constructing a compressor for reading, it uses the dictionary written in the hidden compression options in CompressionInfo.db. And it keeps dictionaries deduplicated, so that each unique live dictionary blob has only one instance in memory, shared across shards.
- We teach the relevant `lz4` and `zstd` compressor wrappers about the dictionaries.
- We add a HTTP API call which samples pieces of the given table (i.e. the Data.db files) from across the cluster, trains a dictionary on it, and publishes it via `system.dicts` as the new current dictionary for that table. (And we add some RPC verbs to support that).

Known things to revisit before merging:
- [ ] There was some discussion about some extra config knobs which aren't added yet. (E.g. an option to disable writing new dict-compressed SSTables altogether).
- [ ] I just stuffed some things into `storage_service` (in particular, the HTTP API call and the RPC verbs, which are things that can't be easily moved later) without thinking much about it, just to get something done. Perhaps they should be elsewhere.
- [ ] I didn't think much about the signatures of the API call and the RPC verbs. We should think twice about them, and about possible needs to extend them later.
- [ ] The callback for `system.dicts` updates is defined in `scylla_main`. That's probably a bad place to keep details of `system.dicts` logic.
- [ ] There is only one basic integration test included. Maybe some parts deserve some more involved testing.
- [ ] The factory currently keeps one dictionary instance on the entire node, but we probably want one copy per NUMA node. I didn't do that because exposing NUMA knowledge to Scylla seems to require some changes in Seastar first.
- [ ] I didn't check the PR for bisectability.

New feature, no backporting involved.

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/15563/)
:white_check_mark: - [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/15563/)
**The following new/updated tests ran 100 times for each mode:**
:small_blue_diamond: boost/cql_query_test
:small_blue_diamond: boost/sstable_3_x_test
:small_blue_diamond: boost/sstable_compaction_test
:small_blue_diamond: boost/sstable_datafile_test
:small_blue_diamond: boost/sstable_test
:small_blue_diamond: topology_custom/test_sstable_compression_dictionaries
:white_check_mark: - [dtest with tablets](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-tablets/3644/)
:white_check_mark: - [dtest with consistent topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/6762/)
:white_check_mark: - [dtest with gossip topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/14786/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/15563/)
#### Build Details:
- Duration: 5 hr 37 min
- Builder: spider2.cloudius-systems.com

avikivity commented:
[ ] user-facing documentation

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/15586/)
:white_check_mark: - [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/15586/)
**The following new/updated tests ran 100 times for each mode:**
:small_blue_diamond: boost/cql_query_test
:small_blue_diamond: boost/sstable_3_x_test
:small_blue_diamond: boost/sstable_compaction_test
:small_blue_diamond: boost/sstable_datafile_test
:small_blue_diamond: boost/sstable_test
:small_blue_diamond: topology_custom/test_sstable_compression_dictionaries
:white_check_mark: - [dtest with consistent topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/6773/)
:white_check_mark: - [dtest with tablets](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-tablets/3656/)
:white_check_mark: - [dtest with gossip topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/14797/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/15586/)
#### Build Details:
- Duration: 5 hr 18 min
- Builder: spider6.cloudius-systems.com

scylladb-promoter commented:
#### :red_circle: CI State: FAILURE

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/15660/)
:x: - [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/15660/)
**The following new/updated tests ran 100 times for each mode:**
:small_blue_diamond: boost/cql_query_test
:small_blue_diamond: boost/sstable_3_x_test
:small_blue_diamond: boost/sstable_compaction_test
:small_blue_diamond: boost/sstable_datafile_test
:small_blue_diamond: boost/sstable_test
:small_blue_diamond: topology_custom/test_sstable_compression_dictionaries
#### Failed Tests (3/424):
- [test_retrain_dict.debug.1](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/15660/testReport/junit/topology_custom/test_sstable_compression_dictionaries/test_retrain_dict_debug_1) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_retrain_dict.debug.1)
- [test_retrain_dict.dev.1](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/15660/testReport/junit/topology_custom/test_sstable_compression_dictionaries/test_retrain_dict_dev_1) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_retrain_dict.dev.1)
- [test_retrain_dict.release.1](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/15660/testReport/junit/topology_custom/test_sstable_compression_dictionaries/test_retrain_dict_release_1) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_retrain_dict.release.1)
#### Build Details:
- Duration: 53 min
- Builder: spider7.cloudius-systems.com

scylladb-promoter commented:
#### :red_circle: CI State: FAILURE

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/15661/)
:white_check_mark: - [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/15661/)
**The following new/updated tests ran 100 times for each mode:**
:small_blue_diamond: boost/cql_query_test
:small_blue_diamond: boost/sstable_3_x_test
:small_blue_diamond: boost/sstable_compaction_test
:small_blue_diamond: boost/sstable_datafile_test
:small_blue_diamond: boost/sstable_test
:small_blue_diamond: topology_custom/test_sstable_compression_dictionaries
:white_check_mark: - [dtest with tablets](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-tablets/3713/)
:white_check_mark: - [dtest with gossip topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/14852/)
:x: - [dtest with consistent topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/6829/)
:x: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/15661/)
#### Failed Tests (2/49162):
- [test_delete_elements_from_a_set](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/15661/testReport/junit/alternator_tests/TesterAlternator/Tests___dtest_with_consistent_topology_changes___test_delete_elements_from_a_set) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_delete_elements_from_a_set)
- [test_topology_ops_encrypted[True].debug.3](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/15661/testReport/junit/topology_custom/test_topology_ops_encrypted/Tests___Unit_Tests___test_topology_ops_encrypted_True__debug_3) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_topology_ops_encrypted[True].debug.3)
#### Build Details:
- Duration: 5 hr 17 min
- Builder: spider7.cloudius-systems.com

scylladb-promoter commented:
#### :red_circle: CI State: FAILURE

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/16191/)
:x: - [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/16191/)
**The following new/updated tests ran 100 times for each mode:**
:small_blue_diamond: boost/cql_query_test
:small_blue_diamond: boost/sstable_3_x_test
:small_blue_diamond: boost/sstable_compaction_test
:small_blue_diamond: boost/sstable_datafile_test
:small_blue_diamond: boost/sstable_test
:small_blue_diamond: cluster/test_sstable_compression_dictionaries_autotrain
:small_blue_diamond: cluster/test_sstable_compression_dictionaries_basic
:small_blue_diamond: cluster/test_sstable_compression_dictionaries_upgrade
#### Failed Tests (5/439):
- [test_upgrade_and_rollback.debug.1](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/16191/testReport/junit/cluster/test_sstable_compression_dictionaries_upgrade/test_upgrade_and_rollback_debug_1) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_upgrade_and_rollback.debug.1)
- [test_upgrade_and_rollback.dev.1](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/16191/testReport/junit/cluster/test_sstable_compression_dictionaries_upgrade/test_upgrade_and_rollback_dev_1) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_upgrade_and_rollback.dev.1)
- [test_upgrade_and_rollback.release.1](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/16191/testReport/junit/cluster/test_sstable_compression_dictionaries_upgrade/test_upgrade_and_rollback_release_1) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_upgrade_and_rollback.release.1)
- [test_describe_simple_schema](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/16191/testReport/junit/boost/cql_query_test/test_describe_simple_schema) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_describe_simple_schema)
- [test_describe_view_schema](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/16191/testReport/junit/boost/cql_query_test/test_describe_view_schema) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_describe_view_schema)
#### Build Details:
- Duration: 5 hr 11 min
- Builder: spider8.cloudius-systems.com

scylladb-promoter commented:
#### :red_circle: CI State: FAILURE

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/16219/)
:x: - [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/16219/)
**The following new/updated tests ran 100 times for each mode:**
:small_blue_diamond: boost/cql_query_test
:small_blue_diamond: boost/sstable_3_x_test
:small_blue_diamond: boost/sstable_compaction_test
:small_blue_diamond: boost/sstable_datafile_test
:small_blue_diamond: boost/sstable_test
:small_blue_diamond: cluster/test_sstable_compression_dictionaries_autotrain
:small_blue_diamond: cluster/test_sstable_compression_dictionaries_basic
:small_blue_diamond: cluster/test_sstable_compression_dictionaries_upgrade
#### Failed Tests (1/878):
- [test_sstable_compression_dictionaries_enable_writing.debug.1](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/16219/testReport/junit/cluster/test_sstable_compression_dictionaries_basic/test_sstable_compression_dictionaries_enable_writing_debug_1) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_sstable_compression_dictionaries_enable_writing.debug.1)
#### Build Details:
- Duration: 1 hr 27 min
- Builder: i-048a31f1a0cf3c8bd (m5d.16xlarge)

scylladb-promoter commented:
#### :red_circle: CI State: FAILURE

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/16229/)
:white_check_mark: - [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/16229/)
**The following new/updated tests ran 100 times for each mode:**
:small_blue_diamond: boost/cql_query_test
:small_blue_diamond: boost/sstable_3_x_test
:small_blue_diamond: boost/sstable_compaction_test
:small_blue_diamond: boost/sstable_datafile_test
:small_blue_diamond: boost/sstable_test
:small_blue_diamond: cluster/test_sstable_compression_dictionaries_autotrain
:small_blue_diamond: cluster/test_sstable_compression_dictionaries_basic
:small_blue_diamond: cluster/test_sstable_compression_dictionaries_upgrade
:x: - [dtest with gossip topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/15359/)
:x: - [dtest with tablets](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-tablets/4190/)
:x: - [dtest with consistent topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/7305/)
:x: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/16229/)
#### Failed Tests (134/20472):
- [test_replace_nonexistent_node[rbo_disabled]](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/16229/testReport/junit/replace_address_test/TestReplaceAddress/Tests___dtest_with_gossip_topology_changes___test_replace_nonexistent_node_rbo_disabled_) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_replace_nonexistent_node[rbo_disabled])
- [test_decommissioned_node_cant_rejoin](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/16229/testReport/junit/topology_test/TestTopology/Tests___dtest_with_gossip_topology_changes___test_decommissioned_node_cant_rejoin) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_decommissioned_node_cant_rejoin)
- [test_replace_nonexistent_node[rbo_enabled]](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/16229/testReport/junit/replace_address_test/TestReplaceAddress/Tests___dtest_with_gossip_topology_changes___test_replace_nonexistent_node_rbo_enabled_) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_replace_nonexistent_node[rbo_enabled])
- [test_removenode_rejoin](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/16229/testReport/junit/update_cluster_layout_tests/TestUpdateClusterLayout/Tests___dtest_with_gossip_topology_changes___test_removenode_rejoin) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_removenode_rejoin)
- [test_removenode_rejoin](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/16229/testReport/junit/update_cluster_layout_tests/TestUpdateClusterLayout/Tests___dtest_with_gossip_topology_changes___test_removenode_rejoin_2) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_removenode_rejoin)
- [test_replace_error_with_different_dc_or_rack](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/16229/testReport/junit/repair_based_node_operations_test/TestRepairBasedNodeOperations/Tests___dtest_with_gossip_topology_changes___test_replace_error_with_different_dc_or_rack) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_replace_error_with_different_dc_or_rack)
- [test_replace_active_node[rbo_enabled]](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/16229/testReport/junit/replace_address_test/TestReplaceAddress/Tests___dtest_with_gossip_topology_changes___test_replace_active_node_rbo_enabled_) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_replace_active_node[rbo_enabled])
- [test_replace_active_node[rbo_enabled]](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/16229/testReport/junit/replace_address_test/TestReplaceAddress/Tests___dtest_with_gossip_topology_changes___test_replace_active_node_rbo_enabled__2) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_replace_active_node[rbo_enabled])
- [test_replace_active_node[rbo_disabled]](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/16229/testReport/junit/replace_address_test/TestReplaceAddress/Tests___dtest_with_gossip_topology_changes___test_replace_active_node_rbo_disabled_) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_replace_active_node[rbo_disabled])
- [test_replace_active_node[rbo_disabled]](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/16229/testReport/junit/replace_address_test/TestReplaceAddress/Tests___dtest_with_gossip_topology_changes___test_replace_active_node_rbo_disabled__2) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_replace_active_node[rbo_disabled])
#### Build Details:
- Duration: 5 hr 26 min
- Builder: i-042be54c3b62f0d9b (m5d.16xlarge)

avikivity commented:
Is this ready for review?

michoecho commented:
> Is this ready for review?

Yes.

raphaelsc commented on replica/database.cc:
```diff
@@ -3184,4 +3184,112 @@ database::on_effective_service_levels_cache_reloaded() {
+            throw std::runtime_error(fmt::format("sample_data_files: table {} does not exist", id));
+        }
+        local_state->schema = t->schema();
+        local_state->snapshot = co_await t->take_sstable_set_snapshot();
```
I think you're fine with just using table::get_sstables(), take_sstable_set_snapshot() is more expensive, since it opens a file desc for every sstable component, and only needed when you need to find file descs for all components that might have been deleted already. You don't seem to have this requirement since you only want to read chunks, and all data is available through sstable object.

michoecho replied:
True.

I will change `take_sstable_set_snapshot()` to copy only the `shared_sstable` objects without opening other components.

(I have to make a copy because AFAICS `table::get_sstables()` returns a shared pointer to a set that can change under my feet if I don't copy it).

michoecho replied:
Done.

raphaelsc commented on replica/database.cc:
```diff
@@ -3184,4 +3184,112 @@ database::on_effective_service_levels_cache_reloaded() {
+            auto permit = co_await local_db._system_read_concurrency_sem.obtain_permit(
+                local_state->schema, "sample_data_files", chunk_size, no_timeout, nullptr);
+            auto [sst, offset] = get_next_chunk();
+            auto buf = co_await sst->data_read(offset, chunk_size, permit);
```
this might enable unnecessary read ahead since data_read() unconditionally uses read ahead of 4 (although it should be reduced to 1 by file input stream even though slow start is not enabled due to lack of history with data_read()) .
Read ahead of 1 means emitting an additional read that will be discarded, so unnecessary read ampl.

Another user of sstable::data_read() is reversed partition read, and it also seems to be affected (it could actually benefit from reversed read ahead but it's not supported AFAIK). Since data_read() is not an actual streaming reader, just consumes a range in one go, I think we should just patch it to disable read ahead altogether.

michoecho replied:
> this might enable unnecessary read ahead

No, because we are passing a finite length (`chunk_size`), and the file stream won't read beyond that length, regardless of what the read ahead is.

I.e. if I ask Seastar to read a file stream from the range [0x1000, 0x2000) of a file with a read ahead of 0x100000, Seastar's is supposed to read [0x1000, 0x2000), not [0x1000, 0x100000].

raphaelsc commented on service/storage_service.cc:
```diff
@@ -4780,6 +4780,77 @@ future<> storage_service::wait_for_topology_not_busy() {
+        chunks_per_host,
+        [&] (std::pair<locator::host_id, uint64_t> h_s) -> future<utils::chunked_vector<bytes>> {
+            const auto& [h, sz] = h_s;
+            return ser::storage_service_rpc_verbs::send_sample_sstables(&ms, h, t, chunk_size, sz);
```
With tablets, some hosts might not have a single tablet of a given table at some point. So maybe it makes sense to skip the rpc call if a host was requested 0 samples, but since training is not frequent, maybe it's irrelevant. 

Very nice you made sample count proportional to the volume on a particular node. This is important for heterogenous setups.

raphaelsc commented on service/storage_service.cc:
```diff
@@ -4784,6 +4784,59 @@ semaphore& storage_service::get_do_sample_sstables_concurrency_limiter() {
+
+future<std::vector<std::byte>> storage_service::train_dict(utils::chunked_vector<bytes> sample) {
+    std::vector<std::vector<std::byte>> tmp;
+    for (const auto& s : sample) {
```
tmp.reserve()

michoecho replied:
Sure, will add.

raphaelsc commented on dict_autotrainer.cc:
```diff
@@ -0,0 +1,160 @@
+        }
+        auto params = s->get_compressor_params();
+        auto ticket = get_units(_ss.get_do_sample_sstables_concurrency_limiter(), 1);
+        auto n_chunks = std::min<uint64_t>(4096, 128*1024*1024 / params.chunk_length());
```
Why 128M? Shouldn't it be some percentage instead of shard's memory? 128M should be ~1% of a i4i cpu, but can be higher on others. I understand you're looking for bigger directionaries for yielding better results. Some comments here explaining why you picked 32M for minimum data size and 128M for sample size could be good for one in the future figuring out how all this works.

If estimated size is 64M, and n_chunks = 4096 (128M / 32k), what happens in practice? redundant samples likely? Maybe instead of constant 128M, use min of 128M and estimated size (rounded down to nearest power of two?) Maybe just increase min_dataset_size to match total chunk size?

michoecho replied:
> Why 128M? Shouldn't it be some percentage instead of shard's memory?

Yes, maybe it should. I'll change it to 1% of total memory.

> Some comments here explaining why you picked 32M for minimum data size and 128M for sample size could be good for one in the future figuring out how all this works.

True, I will add comments. Also, I'll make the minimum data size configurable. (Because tests want it to be smaller than than "real-life" reasonable default).

> If estimated size is 64M, and n_chunks = 4096 (128M / 32k), what happens in practice? redundant samples likely?

In practice, the nodes we are requesting the chunks from will respond with less chunks than requested. (This can't be avoided in general because they might have 0 chunks by the time they receive the request).

Redundant samples would be fine too, mathematically there's nothing wrong with that.

In general, we don't care much about the case when the dataset size is too small to take a full sample. The minimum dataset size condition should prevent that, and if this is some esoteric workload where it doesn't do that, we are fine if the sample is suboptimal. (As long as nothing crashes).

michoecho replied:
Changed the limit to 2% of total memory.

raphaelsc commented on dict_autotrainer.cc:
```diff
@@ -0,0 +1,160 @@
+        }
+        auto ratio_before = co_await try_one_compression_config(_ss.get_database().get_user_sstables_manager().get_compressor_factory(), s, params, validation_sample);
+        auto ratio_after = co_await try_one_compression_config(dict, s, params, validation_sample);
+        constexpr float threshold_fraction = 0.95;
```
this is another place where it's worth a comment to explain why we should publish a new dict that was estimated to be 5% worse than current one. Is it because we think it might yield a better output than current one and that the sample is not sufficient for saying its actual efficiency? Is there any math backing up this decision?

michoecho replied:
True, I'll add a comment to this constant.

> Is it because we think it might yield a better output than current one and that the sample is not sufficient for saying its actual efficiency?

It's because we don't want to train new dictionaries (to avoid general noise on all levels) if they wouldn't be *significantly* better than the current one. So you need some arbitrary criterion for *significantly better*. And the arbitrary criterion here is that the expected compression ratio would decrease by 5%. There's no math behind this constant, and correctness isn't supposed to depend on this constant.

michoecho replied:
Added the comments.

raphaelsc commented on service/storage_service.cc:
```diff
@@ -4773,6 +4780,130 @@ future<> storage_service::wait_for_topology_not_busy() {
+                batch.add_mutation(std::move(publish_new_dict), "publish new SSTable compression dictionary");
+                slogger.debug("publish_new_sstable_dict: committing");
+                co_await std::move(batch).commit(group0_client, local_ss.get_abort_source(), {});
+                slogger.debug("publish_new_sstable_dict: finished");
```
Should we have a metric when a new dict is published so we can correlate changes in compression efficiency with it?

raphaelsc replied:
Maybe another metric for compressed ratio of sstables written with a dict. Since we don't care about sstables written without a dict (which AFAIK can happen when memory usage during writing exceeds a budget), for measuring the efficiency. Maybe another metric for sstable writes that had to fallback to non-dict compression.

michoecho replied:
Note: I didn't apply this suggestion in v3. I'm not sure what they should look like, if they are the right tool at all.

For example, if the metric is global (shared by all tables), then it might stop being a good tool for measuring the efficiency as soon as more than one table exists.

Anyway, we can add more metrics anytime.

avikivity commented on compress.cc:
```diff
@@ -113,6 +108,22 @@ compression_parameters::compression_parameters(compressor_ptr c)
+        case algorithm::deflate: return "DeflateCompressor";
+        case algorithm::snappy: return "SnappyCompressor";
+        case algorithm::zstd: return "ZstdCompressor";
+        case algorithm::none: abort();
```
It's too aggressive to kill a server because of some bug that no one noticed, better on_internal_error() (that only throws in production).

michoecho replied:
Done.

avikivity commented on compress.cc:
```diff
@@ -113,6 +108,22 @@ compression_parameters::compression_parameters(compressor_ptr c)
+        case algorithm::snappy: return "SnappyCompressor";
+        case algorithm::zstd: return "ZstdCompressor";
+        case algorithm::none: abort();
+    }
```
And on_internal_error here too.

michoecho replied:
You mean, add `on_internal_error()` after the switch? I'll add `abort()` there instead.

The compiler will complain if there's a path where a function doesn't return. Which means that we can't go past the switch due to a "benign" programmer error. It can only happen if the enum has an illegal value. And that can only be caused by memory corruption. In this situation, `abort()` makes more sense than `on_internal_error()`.

michoecho replied:
Added `abort()` there.

avikivity commented on compress.hh:
```diff
@@ -18,9 +18,14 @@
+        zstd,
+        snappy,
+        deflate,
+        none,
```
Do we in fact need none? We won't instantiate such a compressor, will we?

michoecho replied:
`none` is what's in the schema when no compression is set.
We don't use `optional<compression_parameters>` anywhere, we have `compression_parameters` with `algorithm::none` instead.

avikivity commented on compress.hh:
```diff
@@ -91,28 +99,32 @@ public:
     static const sstring CRC_CHECK_CHANCE;
 private:
-    compressor_ptr _compressor;
+    algorithm _algorithm;
```
I think that std::optional\<algorithm> captures the different between no compression better (but it is really a minor distinction).

avikivity commented on compress.cc:
```diff
@@ -154,6 +156,51 @@ auto zstd_processor::get_algorithm() const -> algorithm {
+    std::map<int, bytes_view> parts;
+    for (const auto& [k, v] : c.options.elements) {
+        auto k_str = sstring(k.value.begin(), k.value.end());
+        if (k_str.starts_with(DICTIONARY_OPTION)) {
```
Please document the format additions in docs/dev.

michoecho replied:
Done.

avikivity commented on compress.cc:
```diff
@@ -15,10 +15,131 @@
+#include "sstables/sstable_compressor_factory.hh"
+#include "gms/feature_service.hh"
+#include "seastar/core/metrics.hh"
+#include "seastar/core/sharded.hh"
```
CI now complains about \\"seastar/.

avikivity commented on compress.cc:
```diff
@@ -15,10 +15,131 @@
+    ZSTD_customMem as_zstd_custommem() & {
+        return ZSTD_customMem{
+            .customAlloc = [] (void* opaque, size_t n) -> void* {
+                auto addr = malloc(n);
```
What if it returns null?

michoecho replied:
When the allocator returns nullptr to zstd, zstd is supposed to fail at whatever it was doing and return an error code indicating an allocation failure.

(And if you are asking about `malloc_usable_size`, it's supposed to return 0 when called on a nullptr).

michoecho replied:
> and return an error code indicating an allocation failure.

And we check the error code and throw an exception.

avikivity replied:
Yes malloc_usable_size. I expect zstd to handle malloc errors. If malloc_usable_size is prepared, all is well.

avikivity commented on compress.cc:
```diff
@@ -531,3 +652,391 @@ size_t snappy_processor::compress_max_size(size_t input_len) const {
+// To make coordination work without resorting to std::mutex and such, dicts have owner shards,
+// (and are borrowed by foreign shared pointers) and all requests for a given dict ID go through its owner.
+// (Note: this shouldn't pose a performance problem because a dict is only requested once per an opening of an SSTable).
+// (Note: at the moment of this writing, one shard owns all. Later we can spread the ownership. (E.g. shard it by dict hash)).
```
Does "Later" mean a later patch?

avikivity replied:
(in this series)

avikivity replied:
It's okay to do it after the series, but let's do it quickly. I can't predict how much NUMA will affect it and testing will be very hard.

michoecho replied:
Do we even run any tests with NUMA machines?

avikivity commented on compress.cc:
```diff
@@ -531,3 +652,391 @@ size_t snappy_processor::compress_max_size(size_t input_len) const {
+
+zstd_ddict::~zstd_ddict() {
+    if (_owner) {
+        disown();
```
Is this disown() stuff a reimplementation of weak_ptr?

michoecho replied:
Yes, it is.

I had `weak_ptr` in mind and didn't use it for some conscious reason. But now that I look at the code, I don't see the reason anymore.

Will modify.

michoecho replied:
Done.

avikivity commented on replica/database.cc:
```diff
@@ -3184,4 +3184,112 @@ database::on_effective_service_levels_cache_reloaded() {
+        uint64_t local_chunks;
+        uint64_t cross_shard_offset;
+    };
+    std::vector<foreign_ptr<std::unique_ptr<state_by_shard>>> state(smp::count);
```
sharded<>?

michoecho replied:
Right, that would make sense. Will apply the suggestion.

michoecho replied:
Done.

avikivity commented on replica/database.cc:
```diff
@@ -3184,4 +3184,112 @@ database::on_effective_service_levels_cache_reloaded() {
+    };
+    std::vector<foreign_ptr<std::unique_ptr<state_by_shard>>> state(smp::count);
+
+    co_await container().invoke_on_all(coroutine::lambda([&] (auto& local_db) -> future<> {
```
A sharded<> thing can be passed as a parameter to invoke_on_all and auto-unwrapped.

michoecho replied:
Done.

avikivity commented on replica/database.cc:
```diff
@@ -3184,4 +3184,112 @@ database::on_effective_service_levels_cache_reloaded() {
+        int concurrency_limit = 10;
+        co_await max_concurrent_for_each(std::views::iota(uint64_t(0), local_chosen_chunks), concurrency_limit, coroutine::lambda([&] (int) -> future<> {
+            auto permit = co_await local_db._system_read_concurrency_sem.obtain_permit(
+                local_state->schema, "sample_data_files", chunk_size, no_timeout, nullptr);
```
This isn't a regular system ("regular system") read and shouldn't block other system reads. It's executing in the maintenance group and is therefore not interfering with other reads.

avikivity replied:
Ah, but I see you need the permit for calling data_read(). I think it's okay.

avikivity commented on idl/storage_service.idl.hh:
```diff
@@ -77,4 +77,7 @@ verb [[cancellable]] tablet_cleanup (raft::server_id dst_id, locator::global_tab
 verb [[cancellable]] table_load_stats (raft::server_id dst_id) -> locator::load_stats;
 verb [[cancellable]] tablet_repair(raft::server_id dst_id, locator::global_tablet_id) -> service::tablet_operation_repair_result;
+verb [[]] estimate_sstable_volume(table_id table) -> uint64_t;
+verb [[]] sample_sstables(table_id table, uint64_t chunk_size, uint64_t n_chunks) -> utils::chunked_vector<bytes>;
```
It's okay to throw these here.

avikivity commented on service/storage_service.cc:
```diff
@@ -4784,6 +4784,59 @@ semaphore& storage_service::get_do_sample_sstables_concurrency_limiter() {
+                auto batch = service::group0_batch(co_await group0_client.start_operation(local_ss.get_abort_source()));
+                auto write_ts = batch.write_timestamp();
+                auto new_dict_ts = db_clock::now();
+                auto data = bytes(reinterpret_cast<const bytes::value_type*>(dict.data()), dict.size());
```
Can a dictionary exceed 128k?

avikivity replied:
I don't think we need bigger dictionaries either, I'm just checking if I need to prepare for more stall reports or not.

avikivity commented on dict_autotrainer.cc:
```diff
@@ -0,0 +1,160 @@
+            auto tick_period = std::chrono::duration<float>(_cfg.tick_period_in_seconds());
+            auto tick_period_ms = std::chrono::duration_cast<std::chrono::milliseconds>(tick_period);
+            co_await seastar::sleep_abortable(tick_period_ms, _as);
+            if (_ss.is_raft_leader() && _ss.get_feature_service().sstable_compression_dicts) {
```
I don't think this is reliable. Someone might think they are the leader but only discover they're not after trying to commit something. What happens if two leaders clash?

I'm not sure if this is the right way to integrate. Maybe better is to hook it into the topology change coordinator. @patjed41 for opinions on the proper raft way.

michoecho replied:
The leader check is just there to avoid noise. Everything should work correctly even if every node is publishing dicts, but it just may cause more dicts to be flying around than necessary.

The dict is just a recommendation. If two leaders clash, in the worst case some node will use a "short-lived" dict (which is immediately superseded by another one) for an sstable, and now it will have to keep two dicts in memory instead of one, while that sstable is alive.

michoecho replied:
The "right way" would be to have some sort of a distributed cron job. Which, I guess, topology coordinator could be, but we are already stuffing too much topology-unrelated logic into the critical Raft places.

michoecho replied:
Using *group0* for this is icky in the first place. Non-critical things like this should probably be in their own group.

avikivity commented on api/api-doc/storage_service.json:
```diff
@@ -3059,6 +3059,41 @@
          ]
       },
+      {
+         "path":"/storage_service/estimate_compression_ratios",
```
This is fine too.

avikivity commented on test/pylib/scylla_cluster.py:
```diff
@@ -1776,6 +1777,45 @@ async def _server_get_process_status(self, request: aiohttp.web.Request) -> Opti
         return self.cluster.server_get_process_status(server_id)
 
+async def get_scylla_2025_1_executable() -> pathlib.Path:
+    async def run_process(cmd, **kwargs):
```
Let's push this out of the series. It requires a little more review and has a completely different set of requirements. For example, how to we update releases as patch releases advance and as we generate new major releases.

I'm not against it, but it's out of place here.

avikivity commented on test/pylib/scylla_cluster.py:
```diff
@@ -1776,6 +1777,45 @@ async def _server_get_process_status(self, request: aiohttp.web.Request) -> Opti
+        await proc.communicate()
+        assert proc.returncode == 0
+
+    url = 'https://s3.amazonaws.com/downloads.scylladb.com/downloads/scylla/relocatable/scylladb-2025.1/scylla-2025.1.0~rc3-0.20250223.aa5cb15166d3.x86_64.tar.gz'
```
`https://downloads....` should work just as well and is more future-proof.

avikivity replied:
Should we test debug mode against debug mode of older versions? Seems symmetrical and we do want to see if we invoke undefined behavior in order versions.

michoecho replied:
Will address this in the separate series for the upgrade test.

avikivity commented on test.py:
```diff
@@ -195,6 +195,9 @@ def parse_cmd_line() -> argparse.Namespace:
                         help="Specific byte limit for failure injection (random by default)")
+    parser.add_argument('--run-internet-dependent-tests',
+                        dest="run_internet_dependent_tests", action="store_true", default=False,
+                        help="Run tests which need to download something from the internet.")
```
Tests should default to True. Most people do have internet.

The reason I dislike Internet dependency is not because I don't have an internet connection, it's because the tests can break when artifacts are changed or removed.

michoecho replied:
Done.

avikivity commented on test/pylib/scylla_cluster.py:
```diff
@@ -1776,6 +1777,45 @@ async def _server_get_process_status(self, request: aiohttp.web.Request) -> Opti
+    unpack_dir = cache_dir/"unpacked"
+    install_dir = cache_dir/"installed"
+
+    if not install_dir.exists():
```
Very thoughtful

michoecho commented:
> Looks good.
> 
>     * please add a soak test report. That can be any multi-hour, multi-node S-C-T test adjusted so that training is enabled (with at least 2/hour frequency). Preferably one with multiple tables and/or views etc. It doesn't need to show improvements in compression, just that the overall web of RPCs and stuff works in a more realistic setting.
> 
>     * please add a demonstration report. This can be a one-off test where you add compressible data, show compression ratio before and after, list system.dicts state. This is so that if someone wants to tinker with the code, they have a blueprint of how to test it in favorable circumstances. Should have at least 3 nodes with 2 shards each, and (let's say) 20 GB total data.
> 
>     * user-facing documentation
> 
>     * introduce the automated upgrade/downgrade test separately (before or after this series). If after, you can substitute an upgrade/downgrade test report in the cover letter. I like the idea, but it has to be reviewed by the people more familiar with testing.

Ack.

michoecho commented:
v3:
- Addressed review comments.

scylladb-promoter commented:
#### :red_circle: CI State: FAILURE

:white_check_mark: - [Framework test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/16516/)
:x: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/16516/)
#### Build Failure:
```
[2025-03-31T21:51:23.568Z] FAILED: build/dev/sstables/writer.hh.o 
[2025-03-31T21:51:23.568Z] clang++ -MD -MT build/dev/sstables/writer.hh.o -MF build/dev/sstables/writer.hh.o.d -std=gnu++23 -I/jenkins/workspace/scylla-master/scylla-ci/scylla/seastar/include -I/jenkins/workspace/scylla-master/scylla-ci/scylla/build/dev/seastar/gen/include -Werror=unused-result -fstack-clash-protection -DSEASTAR_API_LEVEL=7 -DSEASTAR_BUILD_SHARED_LIBS -DSEASTAR_SSTRING -DSEASTAR_ENABLE_ALLOC_FAILURE_INJECTION -DSEASTAR_LOGGER_COMPILE_TIME_FMT -DSEASTAR_SCHEDULING_GROUPS_COUNT=19 -DSEASTAR_LOGGER_TYPE_STDOUT -DSEASTAR_TYPE_ERASE_MORE -DBOOST_PROGRAM_OPTIONS_NO_LIB -DBOOST_PROGRAM_OPTIONS_DYN_LINK -DBOOST_THREAD_NO_LIB -DBOOST_THREAD_DYN_LINK -DFMT_SHARED -I/usr/include/p11-kit-1 -std=gnu++23 -g -ffile-prefix-map=/jenkins/workspace/scylla-master/scylla-ci/scylla=. -DSEASTAR_NO_EXCEPTION_HACK -march=westmere -DBOOST_ALL_DYN_LINK    -fvisibility=hidden -Ikmipc/kmipc-2.1.0t-rhel84_64/include -DHAVE_KMIP -isystem abseil -Wall -Werror -Wextra -Wimplicit-fallthrough -Wno-mismatched-tags -Wno-c++11-narrowing -Wno-overloaded-virtual -Wno-unused-parameter -Wno-unsupported-friend -Wno-missing-field-initializers -Wno-deprecated-copy -Wno-enum-constexpr-conversion -Wno-error=deprecated-declarations -DXXH_PRIVATE_API -DSEASTAR_TESTING_MAIN -g -ffile-prefix-map=/jenkins/workspace/scylla-master/scylla-ci/scylla=. -DSEASTAR_NO_EXCEPTION_HACK -march=westmere -DDEVEL -DSEASTAR_ENABLE_ALLOC_FAILURE_INJECTION -DSCYLLA_ENABLE_ERROR_INJECTION -DSCYLLA_ENABLE_PREEMPTION_SOURCE -O2 -DSCYLLA_BUILD_MODE=dev -Xclang -fexperimental-assignment-tracking=disabled -iquote. -iquote build/dev/gen  --include sstables/writer.hh -c -o build/dev/sstables/writer.hh.o build/dev/gen/empty.cc
[2025-03-31T21:51:23.569Z] In file included from <built-in>:1:
[2025-03-31T21:51:23.569Z] ./sstables/writer.hh:555:54: error: no template named set in namespace std
[2025-03-31T21:51:23.569Z]   555 |     const encoding_stats& enc_stats = {}, const std::set<int>& compaction_ancestors = {});
[2025-03-31T21:51:23.569Z]       |                                                 ~~~~~^
[2025-03-31T21:51:23.569Z] 1 error generated.
```
#### Build Details:
- Duration: 41 min
- Builder: spider8.cloudius-systems.com

michoecho commented:
v3:
- Refactored `database::sample_data_files()`:
  - Hopefully clarified the logic.
  - Now returns a vector of `temporary_buffer<char>` instead of a vector of `bytes`, and each buffer has semaphore units attached to it. (So that the units are held until they are sent over the wire, not immediately after `sample_data_files` returns). The `SAMPLE_SSTABLES` RPC also returns `temporary_buffer<char>`
  - Now uses `shared` and some of its helpers instead of a vector of foreign pointers.
- `SAMPLE_SSTABLES` RPC also returns `temporary_buffer<char>` instead of `bytes` now.
- Switched that one `abort()` to `on_internal_error()`.
- Switched `take_sstable_set_snapshot` from the expensive `take_storage_snapshot` to a simpler copy of `shared_sstable` objects.
- Added `sstable_compression_dictionaries_min_training_dataset_bytes` and `sstable_compression_dictionaries_min_training_improvement_factor` config knobs (for previously hardcoded constants).
- Changed a logger name from `compressor_factory` to `sstable_compressor_factory`.
- Renamed `dict_autotrainer.{cc,hh}` to `sstable_dict_autotrainer.{cc,hh}`.
- Rewrote the pointers linking the dictonaries and the factory from a custom `disown()` code to `seastar::weak_ptr`.
- Added several debug logs.
- Added a small logging tweak in `fmt_hex` in `bytes.hh`.
- Split the upgrade test out, because it artifact-consuming tests deserve a separate discussion. Will post it separately after this series. (Since it already includes a test for this feature).
- Added a dev doc at `docs/dev/sstable-compression-dicts.md`.
- Added a user-facing doc at `docs/operating-scylla/procedures/config-change/sstable-dictionary-compression.rst`
- Added comments about the arbitrary constants in the sampling.
- Posted a demonstration report at https://github.com/scylladb/scylla-dev/blob/json_loader/demo.md
- Ran a SCT test with dictionaries enabled and high rate of training, at https://argus.scylladb.com/tests/scylla-cluster-tests/80c429fe-158d-4b9e-87d0-1b3e5bf93312, as a "soak test".
  - Used `longevity-mv-si-4days-streaming-test` as the base. (Picked it because Avi asked for something with views, and this was the only test with views I could find which had some green runs).
  - Truncated it to 7 hours from the usual ~1 day.
  - Scaled down the cluster from from 5xi4i.8xlarge to 4x.i4i.xlarge.
  - Scaled down the dataset proportionally to the cluster size.
  - Didn't manage to scale down the load proportionally, so the cluster was relatively overloaded throughout the test.
  - Enabled debug logs on raft loggers and loggers relevant to this PR.
  - Rate-limited a spammy warning about bad view replicas (this test uses tablets with views without respecting the constraints, so illegal things happen during topology operations). 
  - The first time I ran this test, a node bootstrap failed because a group0 snapshot transfer timed out. That was concerning (since this PR touches that are) but after long deliberation I'm 95% sure it's https://github.com/scylladb/scylladb/issues/21637. (Which makes sense because the leader was overloaded at the time, and there are semaphore timeout logs with relevant `system.` reads visible). It didn't reproduce the second time.
  - I didn't know if this is what Avi wanted, but if not, tough luck. Running and tuning SCT is too hard.
- Rebased

michoecho commented:
I'm done with changes for the moment. You can review the changes in v3.

> please add a soak test report. That can be any multi-hour, multi-node S-C-T test adjusted so that training is enabled (with at least 2/hour frequency). Preferably one with multiple tables and/or views etc. It doesn't need to show improvements in compression, just that the overall web of RPCs and stuff works in a more realistic setting.

The pain.

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

:white_check_mark: - [Framework test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/16517/)
:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/16517/)
:white_check_mark: - [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/16517/)
**The following new/updated tests ran 100 times for each mode:**
:small_blue_diamond: boost/cql_query_test
:small_blue_diamond: boost/sstable_3_x_test
:small_blue_diamond: boost/sstable_compaction_test
:small_blue_diamond: boost/sstable_datafile_test
:small_blue_diamond: boost/sstable_test
:small_blue_diamond: cluster/test_sstable_compression_dictionaries_autotrain
:small_blue_diamond: cluster/test_sstable_compression_dictionaries_basic
:white_check_mark: - [dtest with tablets](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-tablets/4459/)
:white_check_mark: - [dtest with consistent topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/7599/)
:white_check_mark: - [dtest with gossip topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/15644/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/16517/)
#### Build Details:
- Duration: 7 hr 4 min
- Builder: spider3.cloudius-systems.com

michoecho commented:
> Looks good.
> 
>     * please add a soak test report. That can be any multi-hour, multi-node S-C-T test adjusted so that training is enabled (with at least 2/hour frequency). Preferably one with multiple tables and/or views etc. It doesn't need to show improvements in compression, just that the overall web of RPCs and stuff works in a more realistic setting.
> 
>     * please add a demonstration report. This can be a one-off test where you add compressible data, show compression ratio before and after, list system.dicts state. This is so that if someone wants to tinker with the code, they have a blueprint of how to test it in favorable circumstances. Should have at least 3 nodes with 2 shards each, and (let's say) 20 GB total data.
> 
>     * user-facing documentation

@avikivity You can check if what I posted in https://github.com/scylladb/scylladb/pull/23025#issuecomment-2767559245 is good enough.

avikivity commented:
I'm sorry that your S-C-T experience was poor. Please list the UX bugs you filed and I'll make sure (with @roydahan) they get fixed.

Regarding the actual test, I expect -streaming means it scales out and in? If I mentioned views I may have mixed it with streaming since that's what exercises dictionary distribution. I don't care about actual results - just to have no crashes related to the series.

michoecho commented:
> Regarding the actual test, I expect -streaming means it scales out and in? If I mentioned views I may have mixed it with streaming since that's what exercises dictionary distribution. I don't care about actual results - just to have no crashes related to the series.

It's a longevity, which means that the test throws random "disruptions" at the cluster. Both in the `-streaming` and non-`-streaming` variant of the test. Scaling out and in was happening too.

`-streaming` means that it uses streaming instead of RBNO. I'm not sure if this even makes any difference if tablets are used. But file streaming was happening in the test, and didn't crash.

> I'm sorry that your S-C-T experience was poor. Please list the UX bugs you filed and I'll make sure (with @roydahan) they get fixed.

Incidentally, have you used SCT? (And if yes — in what ways?)

avikivity commented:
> Incidentally, have _you_ used SCT? (And if yes — in what ways?)

Yes. My Scylla Summit 2024 demonstration of tablet scale in/out was based on an S-C-T test I ran. I filed about 10 bugs.

avikivity commented on CMakeLists.txt:
```diff
@@ -174,6 +174,7 @@ target_sources(scylla-main
     compress.cc
     converting_mutation_partition_applier.cc
     counters.cc
```
> As of this patch, it works like this:
> every `$tick_period` (15 minutes), if we are the current Raft leader,
> we check for dict-aware tables which have no dict, or a dict older
> than `$retrain_period`.

Let's say we have a good dictionary. Doesn't this mean we'll sample every 15 minutes, only to throw the samples away? 

I think it's too much, for large clusters. Once a day should be enough, data doesn't change that often.

michoecho replied:
> > As of this patch, it works like this:
> > every `$tick_period` (15 minutes), if we are the current Raft leader,
> > we check for dict-aware tables which have no dict, or a dict older
> > than `$retrain_period`.
> 
> Let's say we have a good dictionary. Doesn't this mean we'll sample every 15 minutes, only to throw the samples away?

Hmm, right, yes. It's wasteful.

If we find out that the current dictionary is already good, it probably makes sense to refresh its timestamp (the `timestamp` column, not the internal timestamp), so that the trainer doesn't retry this table again for some time, to minimize the noise. It will be a small change.

avikivity replied:
Yes

michoecho commented:
> > Incidentally, have _you_ used SCT? (And if yes — in what ways?)
> 
> Yes. My Scylla Summit 2024 demonstration of tablet scale in/out was based on an S-C-T test I ran. I filed about 10 bugs.

Okay, so the only objective *bugs* I encountered were:
- SCT helpfully deduplicated my `--logger-log-level=X=debug --logger-log-level=Y=debug` arguments, so only one logger ended up with debug logs enabled.
- SCT didn't apply some configs which are applied in regular runs of the test (e.g. encryption, or views-with-tablets), because — I think, but I'm not sure — when it builds a custom AMI, it doesn't add some AMI tags which are added to the `master` AMIs, and without them it could figure out some fact about the Scylla version, and those configs are conditional on the version.   

But the pains aren't (just) about bugs, but about the greater design. I just don't know how to work effectively with it.

For example, to iterate on a test quickly, I'd like my long-running test framework to have things like:
- Fast, local, minimal-scale runs of the test.
- Checkpoints. (So I can e.g. quickly retry only the failed stage several times, or with more logging enabled, or with a new Scylla build).
- Adjustable scale and load.

But I don't know how to iterate on a test without several hours per try. Or how to scale a test down, because the load is all set up in terms of "x instances of type y" and hardcoded cassandra-stress invocations.

More generally, (and this is only a casual user's perspective, of course), I don't understand what SCT can do that can't be done better with a few hundred lines of python+terraform+ssh. To write tests in SCT, you have to learn a whole new set of layers, and for what?

roydahan commented:
> More generally, (and this is only a casual user's perspective, of course), I don't understand what SCT can do that can't be done better with a few hundred lines of python+terraform+ssh. To write tests in SCT, you have to learn a whole new set of layers, and for what?

@michoecho SCT is exactly this few hundreds of lines of python + ssh, which grew over the years to support requests for features and abilities like you need.
Let's move the above comment to SCT and expand on each of the things you need.
Some were already implemented according to your requests (and Avi's) from last survey, e.g. iterations and continue from checkpoints.

mykaul commented:
@michoecho - I did not see any metric that we expose that provides compression stats. Is there any way to expose it? I just saw total_live_memory_bytes which is more of an internal, not user-facing, metric.

michoecho commented:
> @michoecho - I did not see any metric that we expose that provides compression stats. Is there any way to expose it? I just saw total_live_memory_bytes which is more of an internal, not user-facing, metric.

There are no metrics for this. There weren't any before my PR, and I didn't add one.
What kind of stats would you want?

mykaul commented:
> > @michoecho - I did not see any metric that we expose that provides compression stats. Is there any way to expose it? I just saw total_live_memory_bytes which is more of an internal, not user-facing, metric.
> 
> There are no metrics for this. There weren't any before my PR, and I didn't add any. What kind of stats would you want?

How do we know how well the feature works?
@tzach - any specific metrics? Per sstable is obviously useless (and can be probably done out of band via some scylla sstable command), so I'm not sure - table level?

avikivity commented:
> 
> * Fast, local, minimal-scale runs of the test.


That's my # 1 request as well. Someone with low confidence in their S-C-T skills wants to test the test, iteratively, with fast turnaround and low cost. Only then scale the test to realistic proportions.


ps. anyone know how to write `#1` not as code and not as a reference to an issue?

michoecho commented:
> > > How do we know how well the feature works?
> > 
> > 
> > That's what the added `/storage_service/estimate_compression_ratios` HTTP call is for.
> > How do you want to evaluate this feature via metrics? Do you want to see the aggregated compression ratio for all live sstables? In this case, you would have to perform a major compaction to see the effect on the oldest and largest sstables, and the metric isn't worth more than the overall disk usage per table. Do you want to see the compression ratio for newly-written sstables? In this case you don't know if the ratios you see for fresh sstables are representative for the entire table. You would have to perform a major compaction or wait for weeks/months to judge the change. And you won't be able to know if any long-term changes are due to changes in the dataset or changes in the compression algorithm.
> > I think that the only valid way of evaluating the compression algorithms is to test all of them on the same data at once, which is what the HTTP call does.
> 
> It's not (currently) user-facing though (think of Scylla Cloud).

Yes.

But it's not cool that the only things we can expose to the user are metrics. There's a class of measurements which are too expensive for metrics, and this is one of them.

> > > so I'm not sure - table level?
> > 
> > 
> > If you want compression ratio for the entire table, there's the `/column_family/metrics/compression_ratio/{name}` HTTP call. (Maybe it even works).
> 
> This should be also available via regular metrics, so we can present it in the monitoring dashboard. I see its usefulness for the end user.

Doable. Internally we already have stats for total disk usage of sstable sets, I guess the right way would be to add compressed and uncompressed data component sizes to all places where the total is tracked.

>> There's also nodetool tablestats, but, in its infinite wisdom, it rounds the ratio to one decimal place.
>
> We could fix, no? We could extend it?

I hope we could, but you never know if something isn't going to break.

michoecho commented:
> > * Fast, local, minimal-scale runs of the test.
> 
> That's my # 1 request as well. Someone with low confidence in their S-C-T skills wants to test the test, iteratively, with fast turnaround and low cost. Only then scale the test to realistic proportions.

I like how Avi sometimes spuriously responds to months-old comments which were just bumped by something unrelated.
 
> ps. anyone know how to write `#1` not as code and not as a reference to an issue?

@avikivity Like this: #​1

avikivity commented:
> > > * Fast, local, minimal-scale runs of the test.
> > 
> > 
> > That's my # 1 request as well. Someone with low confidence in their S-C-T skills wants to test the test, iteratively, with fast turnaround and low cost. Only then scale the test to realistic proportions.
> 
> I like how Avi sometimes spuriously responds to months-old comments which were just bumped by something unrelated.

My attention span is equal to the height of the github notification inbox.

> 
> > ps. anyone know how to write `#1` not as code and not as a reference to an issue?
> 
> @avikivity Like this: #​1

Ah, a non-breaking space. Though I imagine it breaks many string algorithms.

michoecho commented:
> My attention span is equal to the height of the github notification inbox.
> 
> > > ps. anyone know how to write `#1` not as code and not as a reference to an issue?
> > 
> > 
> > @avikivity Like this: #​1
> 
> Ah, a non-breaking space.

[Zero-width space](https://en.wikipedia.org/wiki/Zero-width_space). [Non-breaking space](https://en.wikipedia.org/wiki/Non-breaking_space) is something else.

> Though I imagine it breaks many string algorithms.

For what it's worth, it doesn't break Ctrl+F in the browser.

avikivity commented:
> > My attention span is equal to the height of the github notification inbox.
> > > > ps. anyone know how to write `#1` not as code and not as a reference to an issue?
> > > 
> > > 
> > > @avikivity Like this: #​1
> > 
> > 
> > Ah, a non-breaking space.
> 
> [Zero-width space](https://en.wikipedia.org/wiki/Zero-width_space). [Non-breaking space](https://en.wikipedia.org/wiki/Non-breaking_space) is something else.

Right, thanks.

> 
> > Though I imagine it breaks many string algorithms.
> 
> For what it's worth, it doesn't break Ctrl+F in the browser.

Interesting that they thought of it.
